test/e2e_node/device-plugin.go

```diff
@@ -0,0 +1,152 @@
+		}
+		if !checkIfNvidiaGPUsExistOnNode() {
+			Skip("Nvidia GPUs do not exist on the node. Skipping test.")
+		}
+
+		framework.Logf("Cluster is running on COS. Proceeding with test")
+
+		if framework.NumberOfNVIDIAGPUs(getLocalNode(f)) == 0 {
+			return
+		}
+
+		installDriver(f)
+
+		// Wait for Nvidia GPUs to be available on nodes
+		Eventually(func() bool { return framework.NumberOfNVIDIAGPUs(getLocalNode(f)) > 0 },
+			driverInstallTimeout*2, time.Second).Should(BeTrue())
+	})
+
+	It("checks that when Kubelet restarts exclusive GPU assignation to pods is kept.", func() {
+		By("Creating one GPU pod on a node with at least two GPUs")
+		n := getLocalNode(f)
+
+		if framework.NumberOfNVIDIAGPUs(getLocalNode(f)) < 2 {
+			Skip("Not enough GPUs to execute this test (at least two needed)")
+		}
+
+		p1 := f.PodClient().CreateSync(makeCudaPauseImage())
+
+		By("Restarting Kubelet and creating another GPU pod")
+		restartKubelet(f)
+
+		p2 := f.PodClient().CreateSync(makeCudaPauseImage())
+
+		cmd := fmt.Sprintf("exec %s %s nvidia-smi -L", n.Name, p1.Spec.Containers[0].Name)
+		uuid1, _ := framework.RunKubectl(cmd)
+
+		cmd = fmt.Sprintf("exec %s %s nvidia-smi -L", n.Name, p2.Spec.Containers[0].Name)
+		uuid2, _ := framework.RunKubectl(cmd)
+
+		By("Checking that pods got a different GPU")
+		Expect(uuid1).To(Not(Equal(uuid2)))
+
+		f.PodClient().DeleteSync(p1.Name, &metav1.DeleteOptions{}, framework.DefaultPodDeletionTimeout)
+		f.PodClient().DeleteSync(p2.Name, &metav1.DeleteOptions{}, framework.DefaultPodDeletionTimeout)
+	})
+})
+
+func installDriver(f *framework.Framework) {
+	dsYamlUrl := "https://raw.githubusercontent.com/GoogleCloudPlatform/container-engine-accelerators/master/device-plugin-daemonset.yaml"
+
+	ds := framework.DsFromManifest(dsYamlUrl)
+	p := &v1.Pod{
+		Spec: ds.Spec.Template.Spec,
+	}
+
+	f.PodClient().CreateSync(p)
+	framework.Logf("Successfully created NVIDIA driver pod installer. Waiting for drivers to be installed and GPUs to be available in Node Capacity...")
+}
+
+func makeCudaPauseImage() *v1.Pod {
+	podName := testPodNamePrefix + string(uuid.NewUUID())
+
+	return &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{Name: podName},
+		Spec: v1.PodSpec{
+			RestartPolicy: v1.RestartPolicyNever,
+			Containers: []v1.Container{{
+				Name:    "cuda-pause",
+				Image:   "nvidia/cuda",
+				Command: []string{"sleep", "infinity"},
+
+				Resources: v1.ResourceRequirements{
+					Limits:   newDecimalResourceList(framework.NVIDIAGPUResourceName, 1),
+					Requests: newDecimalResourceList(framework.NVIDIAGPUResourceName, 1),
+				},
+			}},
+		},
+	}
+}
+
+func newDecimalResourceList(name v1.ResourceName, quantity int64) v1.ResourceList {
+	return v1.ResourceList{name: *resource.NewQuantity(quantity, resource.DecimalSI)}
+}
+
+// TODO: Find a uniform way to deal with systemctl/initctl/service operations. #34494
+func restartKubelet(f *framework.Framework) {
```

Comment:
Given that we only expect the test to run on cos currently, do we still need to support these operations?

Reply:
Don't have any opinion on this. If I had to choose it doesn't seem like a bad solution to support all three init systems and this function can be reused in other tests (and the function is already written).

What init system does COS use ?

Reply:
@jiayingz "only expect the test to run on cos currently" -- why?

Reply:
Currently we only have a single device plugin implementation, i.e., cos based nvidia gpu device plugin. For 1.8 time frame, it seems the easiest way to test the end-to-end device plugin behavior with this reference implementation. In 1.9, we do hope to make the device plugin e2e test generic by using a stub device plugin implementation. I would also hope we may generalize some of the functions used here into a pkg so that different specific device plugin implementations can leverage it in their own e2e testing in separate repositories.